PKCS15 emulator for StarCOS cards with profile eSign/QES

jozsefd opened the pull request:
The PKCS15 emulator is intended for StarCOS 3.4 and 3.5 cards from BA with a profile which contain eSign/QES and the related CIA applications. As discussed in #1882 the on-card PKCS15 structures are broken, but this emulator does not try to fix the EF(ODF) file, but instead, provides direct access to PINs, certificates and keys. 

Since the signature key on v3.5 is 3072 bits wide, the related card-starcos driver was also updated. During testing PKCS11 and Minidriver several issues were amended as well:
- 3072 bit key support
- card logout v3.x cards
- PIN GET_INFO logged in state support for v3.5 cards (required in PKCS11/Firefox)
- path caching is disabled on Windows, because it may confuse parallel minidriver instances

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] Documentation is added or updated
- [x] New files have a LGPL 2.1 license statement
- [x] PKCS#11 module is tested
- [x] Windows minidriver is tested
- [ ] macOS tokend is tested

Jakuje commented:
Can you rebase the changes on current master so they are mergable? Also all the CI build are failing. Can you check these so it builds?

jozsefd commented:
For sure, I will do it and the double check the CI build.

jozsefd commented:
The Windows build is less restrictive, that's why the proposed changes had CI errors. The Windows Build Instructions wiki seems to be outdated anyway.
Now it should be OK, please check!

frankmorgner commented on src/libopensc/pkcs15-starcos-esign_qes.c:
```diff
@@ -0,0 +1,366 @@
+
+    if (!p15card || ! p15card->card || !p15card->card->ctx) return SC_ERROR_INVALID_ARGUMENTS;
+
+    r = starcos_esign_qes_init(p15card, aid);
```
I suppose that StarCOS is quite flexible regarding the card profile. It looks like this emulator assumes that all 3.4/3.5 cards have the QES profile. And if they have the QES app installed, then it is assumed that the card doesn't have any other applications. Is it possible to differentiate your card profile from others? Most vendors use a specific ATR, for example...

An other problem is that you most likely don't need the generic profile parsing with `sc_pkcs15_bind_synthetic()`. If you add your card to `sc_pkcs15_is_emulation_only()`, then your profile is loaded right away even if `try_emulation_first` is not set in `opensc.conf`...

jozsefd replied:
You are right these cards should have a specific ATR, I can check for them and introduce a STARCOS subtype with this profile.

On the other hand I have tested it with a v3.4 D-Trust card which has completely different profile. In this case the emulator reports WRONG_CARD so even if try_emulation_first is set, the card would work with the PKCS15 info on the card. But if either one of the supported apps are card then it would activate the emulator (in case of PKCS15 or minidriver, the PKCS11 Framework uses a different approach by loading all apps from EF.DIR)

frankmorgner replied:
I don't know what hinders the D-Trust card to be loaded, but please make the emulation profile specific for your card only.

jozsefd commented on src/libopensc/card-starcos.c:
```diff
@@ -280,6 +281,54 @@ static unsigned int starcos_determine_pin_encoding(sc_card_t *card)
+		}
+	}
+	return 0;
+}
```
Yes, but the same "Command Sequence" section of the 3.5 documentation says:
After successful command execution, STARCOS performs the following: 1. Sets the state of the PIN or password to „verified“.
In practice, the empty VERIFY also works on the the v3.5 cards, but it never answers '90 00', but always 'C3 Cx'. This means in Firefox/P11 the login function will not work. 
That's why this workaround is only used for v3.5 cards, but I agree, it would be better to determine the logged_in status by the card, I will double check if it is possible.

dengert replied:
Some PIV like cards have the same problem with "empty VERIFY" and testing for login state. 
`#define CI_VERIFY_LC0_FAIL                  0x00000002U /* VERIFY Lc=0 never returns 90 00 if PIN not needed */` and addresses it by `/* try another method, looking at a protected object this may require adding one of these to NEO */`  Look at the logic starting at line 5281 ` /* Some cards never return 90 00  for SC_PIN_CMD_GET_INFO even if the card state is verified */`.

jozsefd replied:
Unfortunately I did not find any StarCOS 3.x data object or file which could produce the "69 82"/"Security status not satisfied" error besides using a private key. But the card driver itself is unaware of any private keys, this information comes from the higher layers like PKCS15. I keep searching for a solution

jozsefd replied:
I was stuck with this issue a little bit :) 
At the end, I even managed to get a contact from the card manufacturer, but unfortunately they have confirmed: it is not possible to query the password's security status on the card with this Firmware Version of StarCOS 3.5 (IFAIK v3.52). Should I leave or remove the software workaround?

jozsefd commented on src/libopensc/card-starcos.c:
```diff
@@ -826,9 +878,15 @@ static int starcos_select_file(sc_card_t *card,
+#ifdef ENABLE_MINIDRIVER
+		0; // multiple windows minidriver instances may exists, path caching makes no sense
+#else
+		card->cache.valid;
```
Yes, I agree. Do you think it would make sense leave a compile time option for path caching? Or better remove from the code.

frankmorgner replied:
Leave it as is. When we remove this, we will do so throughout the whole code base.

jozsefd commented on src/libopensc/card-starcos.c:
```diff
@@ -840,23 +898,42 @@ static int starcos_select_file(sc_card_t *card,
+
+	if (in_path->aid.len) {
+		if (!pathlen) {
+			memcpy(path, in_path->aid.value, in_path->aid.len);
```
The size of path buffer is SC_MAX_PATH_SIZE and aid.value buffer is SC_MAX_AID_SIZE and both constants are 16 so it should be OK, but I agree, we cannot rely on this match, and the aid.len (or even the path.len) value may be incorrect, so I must check for buffer overrun.

jozsefd commented on src/libopensc/card-starcos.c:
```diff
@@ -840,23 +898,42 @@ static int starcos_select_file(sc_card_t *card,
+				LOG_TEST_RET(card->ctx, r, "Could not select AID!");
+			}
+
+			if (pathtype == SC_PATH_TYPE_DF_NAME) pathtype = SC_PATH_TYPE_FILE_ID;
```
OK. I would rather add then curly brackets too. (I prefer if statements w/o curly brackets as one-liner)

jozsefd commented on src/libopensc/card-starcos.c:
```diff
@@ -1713,17 +1792,37 @@ static int starcos_compute_signature(sc_card_t *card,
+
+			if (card->caps & SC_CARD_CAP_APDU_EXT
+				&& card->reader->max_recv_size > SC_MAX_APDU_BUFFER_SIZE
+				&& outlen > SC_MAX_APDU_BUFFER_SIZE ) {
```
The GET RESPONSE command is not included in the 3.4 nor in the 3.5 manual, so I suppose it is not supported, but I can check it. This patch was intended for the v3.5 cards with 3072 bit RSA key. In this case the answer to COMPUTE DIGITAL SIGNATURE would exceed 256 bytes, and it would only work with extended APDU case 4. 
I will double check if the card works with SC_APDU_FLAGS_CHAINING.

frankmorgner replied:
Chaining is for the command, not the response. I just wanted to make you aware of the problem that some readers may not be able to send/receive your data.

If there is no workaround for 3072 bit RSA with an only short-length-APDU-capable reader, then you should patch the code to assume that the reader is capable of sending extended length, as done in the [CardOS driver](https://github.com/OpenSC/OpenSC/blob/929717b505c8715cc97d0b44c965e09bb785a2d8/src/libopensc/card-cardos.c#L1199-L1254), for example. Unfortunately, the reader's capabilities cannot be reliably detected, so it's better to assume all requirements are met rather than failing without even trying.

frankmorgner replied:
did you have the chance to look at the cardos link?

jozsefd replied:
Sorry for the delay and yes, I have already implemented a function to probe the reader possibilities, similar to the CardOS driver. Unfortunately I was quite busy in the past weeks, hopefully I will find time on the weekend to rebase the branch and test it through.

jozsefd commented on src/libopensc/card-starcos.c:
```diff
@@ -2010,12 +2107,32 @@ static int starcos_card_ctl(sc_card_t *card, unsigned long cmd, void *ptr)
+	int r = SC_ERROR_NOT_SUPPORTED;
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_NORMAL);
+
+	starcos_pin_clear_logged_in(card);
```
This patch is a workaround for an awkward error I encountered during testing: Windows minidriver accepted an empty PIN in the second instance of Outlook/SMIME (well, it happened on Win7 only). The StaCOS docu says:

---
### Deletion of Security States
The security states of the STARCOS smart are deleted as follows:
- During the RESET of the smart card, all security states of the STARCOS smart card are deleted.
- When selecting a DF, the security states of a selected DF and the DFs, which are superior to this DF up to the MF, remain unchanged. All other security states are deleted.
- If an SE is activated in a DF, all security states allocated to this DF are deleted.
- If a key stored in volatile memory is deleted, the security state belong- ing to this key or the negotiation key or the master are deleted.
---

I will check if the above ISO7816 VERIFY option works.

jozsefd replied:
I have double checked this ISO7816 APDU option: it is not in the documentation and does not work for neither v3.4 nor v3.5 StarCOS cards. It produces response 6A 86 - Incorrect parameters P1-P2.

But I found an APDU sequence which could reset the security states:
- select MF
- MSE RESTORE #02 (00 22 F3 02)

The card driver uses security environment #01 so restoring #02 would effectively clear the states. I could prove it for both 3.4 and 3.5 cards with blank VERIFY and with COMPUTE DIGITAL SIGNATURE. I did not find any side effect of this workaround.

jozsefd commented on src/libopensc/pkcs15-starcos-esign_qes.c:
```diff
@@ -0,0 +1,366 @@
+
+static int get_cert_size(sc_card_t * card, sc_path_t * path, size_t * psize) {
+	int r;
+	sc_file_t * file;
```
OK

frankmorgner commented:
what's the status of this pull request? as far as i can see, there are some comments not yet addressed, right?

jozsefd commented:
Sorry for the delay, I tried to solve all issues, and I have proposed solution all for except one: the PIN GET_INFO logged in state of StarCOS 3.5 cards can be solved with a software solution in the driver.

frankmorgner reviewed (changes requested):
OK, we'll wait until you have the time for implementing.

Please also note that the indenting across your code is consistently bad mixing tabs and spaces...

frankmorgner commented on src/libopensc/card-starcos.c:
```diff
@@ -280,6 +281,54 @@ static unsigned int starcos_determine_pin_encoding(sc_card_t *card)
+ * Returns -1 if the pin state cannot be set
+ * Note: pin reference must be greater than 0
+ */
+static int starcos_pin_set_logged_in(sc_card_t *card, unsigned char pin_reference) {
```
I'm not sure how useful it is to cache the PIN verification status at the driver level. There may be other processes interfering with the card, especially in a long running process such as in the Browser.

I think it should be OK to just leave out the pin status query. The upper software layers (PKCS#11, minidriver) should handle this correctly...

jozsefd replied:
Yes, the minidriver can handle this, but some PKCS#11 clients, especially Firefox, cannot "Log in" with StarCOS 3.5: it happens to verify the PIN successfully, but forgets this status immediately and tries to query it. 
This means client side SSL authentication is not possible without this patch.

jozsefd replied:
And you are rigth, sorry for the bad indenting. Seems my development environments unter Windows and under Linux do not match, although both are VS Code. I am goint to fix this as well!

frankmorgner commented:
@jozsefd any update on this? there is only one unresolved conversation and are some cosmetic changes left...

jozsefd commented:
I did run extensive tests with this branch in December, including all possible profile variants of the supported StarCOS cards and with 5 different card readers. Unfortunately, the handling of extended APDUs did not work in some reader/card combinations, I have fixed it anyway. 
I found another issue which is likely unrelated to this PR, but it prevented me to submit the changes: if the test system has multiple card readers attached then the Windows minidriver fails when switching a card between readers. I'd like to analyze and understand this problem before submitting the PR again.

frankmorgner commented:
OK, thanks for the update

frankmorgner commented:
This PR looks good from my perspective. Could you give me an update on the open issue? I'd like to know if we can include this PR in a new release or not...

jozsefd commented:
Sorry for the long delay. I tried to double check and solve the issue in minidriver with multiple card readers and switching the card between the readers. In that case the card works only in the first card reader, the minidriver fails to initialize the card in any other card reader.
I thought it was an error in my PR, but when I backported it to an older OpenSC Version (e.g. 0.19), it turned out that not my PR but the changed PCSC and Minidriver handling caused this issue. I suppose it was introduced with fix of the issue #1874 last year. In fact it could be a problem of the Microsoft Smartcard Subsystem, the analysis of the problem showed the following:
1) card plugged in the second card reader
2) the minidriver CardAcquireContext, CardGetProperty, CardReadFile(cmapfile) calls are successfully completed, CardDeleteContext called
3) the minidriver CardReadFile(cmapfile) gets called with the previous (still valid?) pCardData, 
4) "HANDLES CHANGED" detected, the reader-pcsc tries "Reusing the reader", the SCardGetAttrib(SCARD_ATTR_DEVICE_SYSTEM_NAME_A) reports the first (wrong) card reader: connection fails!

Older versions of OpenSC happened to reinitialize everything in this case, the new version simply fails. I tried several workarounds, e.g.: when I included the reader name in the container names of the cmapfile (HASH(card id + container id + reader system name)), then it worked. I report this as a separate issue, maybe you have a better solution.

frankmorgner commented:
What's the status of this topic, is there anything to do?

jozsefd commented:
> What's the status of this topic, is there anything to do?

This topic is still open. A minidriver built from this branch was deployed in several systems and based on the problems occurred and on the feedback from the card vendor (G&D) I changed the following:
- added support for PSS padding on v3.5
- removed PIN state tracking from starcos driver
- removed the QES application (qualified signing) from the emulator upon comments
- fix select file in starcos driver

It seems, it would make sense to revoke this PR and post a new one. Would it be OK for you, or you'd prefer to apply all changes to this branch?

frankmorgner commented:
Go ahead with what you think works best for you.

jozsefd commented:
> Go ahead with what you think works best for you.

Thank you, then I close this PR and make a new one soon with the actual version of StarCOS eSign profile cards.
